templates: make node compilation optional (#5954)
paritytech/polkadot-sdk: substrate/bin/utils/chain-spec-builder/README.md

runtime functions are called.

### Generate raw storage chain spec using genesis config patch

Patch the runtime's default genesis config with provided `patch.json` and generate raw
storage (`-s`) version of chain spec:

<!-- docify::embed!("tests/test.rs", cmd_create_with_patch_raw)-->

_Note:_ [`GenesisBuilder::build_state`](https://docs.rs/sp-genesis-builder/latest/sp_genesis_builder/trait.GenesisBuilder.html#method.build_state)
runtime function is called.

### Generate raw storage chain spec using full genesis config

Build the chain spec using provided full genesis config json file. No defaults will be used:

<!-- docify::embed!("tests/test.rs", cmd_create_full_raw)-->

_Note_: [`GenesisBuilder::build_state`](https://docs.rs/sp-genesis-builder/latest/sp_genesis_builder/trait.GenesisBuilder.html#method.build_state)
runtime function is called.

### Generate human readable chain spec using provided genesis config patch

<!-- docify::embed!("tests/test.rs", cmd_create_with_patch_plain)-->

### Generate human readable chain spec using provided full genesis config

<!-- docify::embed!("tests/test.rs", cmd_create_full_plain)-->

### Extra tools

> The `chain-spec-builder` provides also some extra utilities: `VerifyCmd`, `ConvertToRawCmd`, `UpdateCodeCmd`.

Reviewer: ```suggestion
The `chain-spec-builder` provides also some extra utilities: [`VerifyCmd`](https://docs.rs/staging-chain-spec-builder/latest/staging_chain_spec_builder/struct.VerifyCmd.html), [`ConvertToRawCmd`](https://docs.rs/staging-chain-spec-builder/latest/staging_chain_spec_builder/struct.ConvertToRawCmd.html), [`UpdateCodeCmd`](https://docs.rs/staging-chain-spec-builder/latest/staging_chain_spec_builder/struct.UpdateCodeCmd.html).
```

Author: [5b16b64](https://github.com/paritytech/polkadot-sdk/pull/5954/commits/5b16b64eed68a0335a8581986d8b140ad0828636)

> # Chain Spec Builder

Reviewer: For me, it makes little sense to copy the part of [other documentation]((https://github.com/paritytech/polkadot-sdk/blob/58fd5ae4ce883f42c360e3ad4a5df7d2258b42fe/substrate/bin/utils/chain-spec-builder/src/lib.rs#L19)) to the readme file. It would be better to move the everything to the readme file and then include it into the module documentation using:
```
#![doc = include_str!("../README.md")]
```

IMO copy is not maintainable.

Reviewer: Also - we could just link existing doc page:
https://docs.rs/staging-chain-spec-builder/latest/staging_chain_spec_builder/
or polkadot-sdk doc (which is the same):
https://paritytech.github.io/polkadot-sdk/master/staging_chain_spec_builder/index.html

Author: Great suggestion! :+1:  I like better the first option because it should look the same for all three doc sources down the line, while maintaining a single file.

Substrate's chain spec builder utility.

> A chain-spec is short for `chain-specification`. See the [sc-chain-spec](https://crates.io/crates/sc-chain-spec)

Reviewer: Referring to docs make more sense:
```suggestion
A chain-spec is short for `chain-specification`. See the [sc-chain-spec](https://docs.rs/sc-chain-spec/latest/sc_chain_spec/)
```

Author: [37e43c0](https://github.com/paritytech/polkadot-sdk/pull/5954/commits/37e43c03090ee56e14c5df4da2c42945e6bcc08e)

> A chain-spec is short for `chain-configuration`. See the [sc-chain-spec](https://crates.io/crates/sc-chain-spec)

Author: [ccad82f](https://github.com/paritytech/polkadot-sdk/pull/5954/commits/ccad82fa37b7ea69df17486b7de4549d393121a0)

for more information.

> **Note**: this binary is analogous to the `build-spec` subcommand, contained in typical Substrate-based nodes.

Author: [ccad82f](https://github.com/paritytech/polkadot-sdk/pull/5954/commits/ccad82fa37b7ea69df17486b7de4549d393121a0)

This particular binary is capable of interacting with [sp-genesis-builder](https://docs.rs/sp-genesis-builder/latest/sp_genesis_builder/index.html)
implementation of any provided runtime allowing to build chain-spec JSON files.

See [ChainSpecBuilderCmd](https://docs.rs/staging-chain-spec-builder/6.0.0/staging_chain_spec_builder/enum.ChainSpecBuilderCmd.html)
for a list of available commands.

## Installation

```bash
cargo install staging-chain-spec-builder

> ```

Reviewer: Actually it would be good to fix the original file, and make this real use case with `#[docify]` test using _command-macro_ as in https://github.com/paritytech/polkadot-sdk/issues/6194.

Author: This was fun. It doesn't look as good as before, but the test docs suggest enough for `chain-spec-builder` binary usage. You can see the required changes here: [24ec943](https://github.com/paritytech/polkadot-sdk/pull/5954/commits/24ec943cccd3abd6e19e495abbec9d7161ce50c3).

Reviewer: Yeah, idea looks super good. I like that we have real commands backed by tests.

To improve rendering (which now does not look well) - maybe we could do sth like this:
```rust
  use cmd_lib::run_cmd;

  macro_rules! exe(
      ( chain-spec-builder $($a:tt)* ) => {
          let bin_path = env!("CARGO_BIN_EXE_chain-spec-builder");
         run_cmd!(
             $bin_path $($a)*
          );
      }
  );

  #[test]
  #[docify::export_content]
  fn it_works() {
     exe!(chain-spec-builder list-preset);
  }    
```

This way the doc would like this:
![image](https://github.com/user-attachments/assets/bbf26185-c2bd-4838-a846-c36077d45246)

Which is a bit more readable. wdyt?

Reviewer: It is actually not strictly related to making the compilation optional, so maybe this (entire re-work) could be taken to some followup.

Reviewer: I was also doing a hackish trick to pass `-c /dev/stdout` to display the output instead writing to file.

Author: > We can also use /dev/null to hide that output.

Might be confusing from an example perspective, but comments might help. How does something like:

```rust
#[test]
#[docify::export_content]
fn create_default() {
	let runtime_path = substrate_test_runtime::WASM_BINARY_PATH.expect("a valid path. qed.");
	// Can be set to any file, default is `chain_spec.json`.
	let output_path = "/dev/null";
	exe!(chain-spec-builder -c $output_path create -r $runtime_path default);
}
```

Reviewer: I was thinking about this direction (rust-pseudo-code):
```rust
  #[docify::export_content]
  fn create_default_doc_call(...) -> serde_json::Value {
     let output = exe!(chain-spec-builder -c /dev/null create  -r runtime.wasm default).stdout_utf8()
     output.to_json()
  }  

#[test]
fn create_default() {
	let runtime_path = substrate_test_runtime::WASM_BINARY_PATH.expect("a valid path. qed.");
        output = create_default_doc_call(...runtiime_path...) ;
        expected_json_from_file = read_file().to_json();
        assert_eq!(ouyput, expected_json_from_file );
}
```

Reviewer: So basically what we put to the doc is _pure cmd line call_ ideally with some "real paths".

And the test checks the output of this _pure call_ using rust. I hope I made it clear - if not let's iterate.

Reviewer: > Using `/dev/stdout` can work but it will pollute the stdout of the tests, which will make searching through the failures harder. We can also use `/dev/null` to hide that output.

It will be captured by macro (and internall by `Command`), so you will not actually see it in output.

Reviewer: can't stop thinking about this, sorry for throwing random thoughts: we could even hide `-c /dev/stdout ` in the helper `exe!` macro. we could do similar thing with `runtime.wasm` file.

Reviewer 2: I like the overall approach of `exe!`, or any of the tools mentioned in https://github.com/paritytech/polkadot-sdk/issues/6194. 

Note that in both cases, you need https://docs.rs/docify/latest/docify/macro.compile_markdown.html as we are in a markdown file, not in `.rs` anymore. I have never used it myself, but it should meet the needs here.

Author: > Note that in both cases, you need https://docs.rs/docify/latest/docify/macro.compile_markdown.html

Yes, that's what I used here: [24ec943](https://github.com/paritytech/polkadot-sdk/pull/5954/commits/24ec943cccd3abd6e19e495abbec9d7161ce50c3).

###  Regarding  https://github.com/paritytech/polkadot-sdk/pull/5954#discussion_r1822551099

Tried a bunch of things to make it work like in the example, but the closest I was able to is here: [231802c](https://github.com/paritytech/polkadot-sdk/pull/5954/commits/231802c0390bc5ef7464e2825c69a140f081db2a).

Example pic:
![Screenshot from 2024-10-30 22-17-39](https://github.com/user-attachments/assets/154ff303-4627-4d5e-b391-2be20f7035e2)

Reasons for why we can't go with the separation between `docify` and `#[test]` for code samples we want to embed into the crate-level docs of a crate:

1. the function exported with `docify` will be compiled as a doc test (runnable with `cargo test` - although the docify docs says that `docify::embed!` - the one used in the README.md - will wrap things with `ignore`, but seems like this isn't the case - might be a bug, haven't investigated much). This results in rust code that executed the commands being compiled outside the integration tests target, so it will not be able to use the `CARGO_BIN_EXE_chain-spec-builder` environment variable, and even not find `exe!` macro (declared in `tests.rs`).

2. The first issue alone (missing `CARGO_BIN_EXE_chain-spec-builder`) requires exporting the entire test with `docify::export` (not `docify::export_content`), so that docs' tests are successful. The second issue (not finding `exe!` macro, defined in `tests.rs`) isn't that easy to fix either - I tried but without luck.

3. EXTRA: not necessarily related to the separation between `docify` and `#[test]`, but considering the aspect of `ignoring` running the rust code present in docs, I thought that we might be able to drop the full `README.md` (single source of docs) and use instead the crate-level lib.rs docs where I would embed things with `docify::embed`, which hopefully will wrap around with `ignore` the embedded rust code, and we'll not run it as docs test. However `docify::export*` doesn't work for crate-level lib.rs docs, only for general public docs (`///`).

This being said, I would say it is not worth investing more effort in making the docs look nicer than they currently look for `staging-chain-spec-builder`. @michalkucharczyk lmk though if you still have outstanding feedback. In the meantime I will focus on the rest of the feedback in the meantime, since we want to merge the PR by this Friday.

> **Note**: `chain-spec-builder` binary is published on [crates.io](https://crates.io) under

Author: [ccad82f](https://github.com/paritytech/polkadot-sdk/pull/5954/commits/ccad82fa37b7ea69df17486b7de4549d393121a0)

> [staging-chain-spec-builder](https://crates.io/crates/staging-chain-spec-builder) due to a name conflict.

Reviewer: ```suggestion
[`staging-chain-spec-builder`](https://crates.io/crates/staging-chain-spec-builder) due to a name conflict.
```

Reviewer: and also all links below...

Author: [37e43c0](https://github.com/paritytech/polkadot-sdk/pull/5954/commits/37e43c03090ee56e14c5df4da2c42945e6bcc08e)

## Usage

### Generate chains-spec using default config from runtime

Query the default genesis config from the provided `runtime.wasm` and use it in the chain
spec.

```bash
chain-spec-builder create -r runtime.wasm default
```

> _Note:_ [GenesisBuilder::get_preset](https://docs.rs/sp-genesis-builder/latest/sp_genesis_builder/trait.GenesisBuilder.html#method.get_preset)

Reviewer: IMO this would look better:
```suggestion
_Note:_ [`GenesisBuilder::get_preset`](https://docs.rs/sp-genesis-builder/latest/sp_genesis_builder/trait.GenesisBuilder.html#method.get_preset)
```

Author: [37e43c0](https://github.com/paritytech/polkadot-sdk/pull/5954/commits/37e43c03090ee56e14c5df4da2c42945e6bcc08e)

### Generate raw storage chain spec using genesis config patch

Patch the runtime's default genesis config with provided `patch.json` and generate raw
storage (`-s`) version of chain spec:

```bash
chain-spec-builder create -s -r runtime.wasm patch patch.json
```

_Note:_ [GenesisBuilder::build_state](https://docs.rs/sp-genesis-builder/latest/sp_genesis_builder/trait.GenesisBuilder.html#method.build_state)
runtime function is called.

### Generate raw storage chain spec using full genesis config

Build the chain spec using provided full genesis config json file. No defaults will be used:

```bash
chain-spec-builder create -s -r runtime.wasm full full-genesis-config.json
```

_Note_: [GenesisBuilder::build_state](https://docs.rs/sp-genesis-builder/latest/sp_genesis_builder/trait.GenesisBuilder.html#method.build_state)
runtime function is called.

### Generate human readable chain spec using provided genesis config patch

```bash
chain-spec-builder create -r runtime.wasm patch patch.json
```

### Generate human readable chain spec using provided full genesis config

```bash
chain-spec-builder create -r runtime.wasm full full-genesis-config.json
```

### Extra tools

> The `chain-spec-builder` provides also some extra utilities: `VerifyCmd`, `ConvertToRawCmd`,

Reviewer 2: I think this page is really good, and generally much needed, given that `ChainSpecBuilder` and its related code is new. 

My nitpick for future would also be for you to consider only covering the most important user flows in the `README`, and leave the rest to `please see --help` for more. 

I am a fan of this principle as it reduces the maintenance of the README. for complex CLIs, copy-pasting the entire CLI spec into the readme will soon become a mess. It is more worthwhile to improve the documentation of the types in Rust that will lead to a more high quality `--help`. 

Feel free to incorporate this here as well, but I will leave it up to you to decide.

Author: I think I am good with how it is currently. 

At this point this `README.md` is the single source of truth for `chain-spec-builder`, with commands that are backed by integration tests. The `README.md` was created based on what we had in `lib.rs` as crate-level docs (which appear in `docs.rs` and `polkadot-sdk-docs`). I am thinking this can stay like this for now until a more opinionated view over it will be proposed.
